cmd/juju/controller/addmodel.go

```diff
@@ -412,98 +420,140 @@ more than one credential is available. List credentials with:
 and then run the add-model command again with the --credential flag.`[1:],
 )
 
-func (c *addModelCommand) maybeUploadCredential(
-	ctx *cmd.Context,
-	cloudClient CloudAPI,
-	cloudTag names.CloudTag,
-	cloudRegion string,
-	cloud jujucloud.Cloud,
-	modelOwner string,
-) (names.CloudCredentialTag, error) {
+type findCredentialParams struct {
+	cloudTag    names.CloudTag
+	cloud       jujucloud.Cloud
+	cloudRegion string
+	modelOwner  string
+}
+
+// findCredential finds a suitable credential to use for the new model.
+// The credential will first be searched for locally and then on the
+// controller. If a credential is found locally then it's value will be
+// returned as the first return value. If it is found on the controller
+// this will be nil as there is no need to upload it in that case.
+func (c *addModelCommand) findCredential(ctx *cmd.Context, cloudClient CloudAPI, p *findCredentialParams) (_ *jujucloud.Credential, _ names.CloudCredentialTag, cloudRegion string, _ error) {
+	if c.CredentialName == "" {
+		return c.findUnspecifiedCredential(ctx, cloudClient, p)
+	}
+	return c.findSpecifiedCredential(ctx, cloudClient, p)
+}
 
+func (c *addModelCommand) findUnspecifiedCredential(ctx *cmd.Context, cloudClient CloudAPI, p *findCredentialParams) (_ *jujucloud.Credential, _ names.CloudCredentialTag, cloudRegion string, _ error) {
+	fail := func(err error) (*jujucloud.Credential, names.CloudCredentialTag, string, error) {
+		return nil, names.CloudCredentialTag{}, "", err
+	}
 	// If the user has not specified a credential, and the cloud advertises
 	// itself as supporting the "empty" auth-type, then return immediately.
-	if c.CredentialName == "" {
-		for _, authType := range cloud.AuthTypes {
-			if authType == jujucloud.EmptyAuthType {
-				return names.CloudCredentialTag{}, nil
-			}
+	for _, authType := range p.cloud.AuthTypes {
+		if authType == jujucloud.EmptyAuthType {
+			return nil, names.CloudCredentialTag{}, p.cloudRegion, nil
 		}
 	}
 
-	// Check if the credential is already in the controller.
-	modelOwnerTag := names.NewUserTag(modelOwner)
-	credentialTags, err := cloudClient.UserCredentials(modelOwnerTag, cloudTag)
+	// No credential has been specified, so see if there is one already on the controller we can use.
+	modelOwnerTag := names.NewUserTag(p.modelOwner)
+	credentialTags, err := cloudClient.UserCredentials(modelOwnerTag, p.cloudTag)
+	if err != nil {
+		return fail(errors.Trace(err))
+	}
+	var credentialTag names.CloudCredentialTag
+	if len(credentialTags) == 1 {
+		credentialTag = credentialTags[0]
+	}
+
+	credential, credentialName, cloudRegion, err := c.findLocalCredential(ctx, p, credentialTag.Name())
+	if (credentialTag != names.CloudCredentialTag{}) && (err == nil || errors.IsNotFound(err)) {
```

Comment:
it would make mode sense to the reader if  (credentialTag != names.CloudCredentialTag{}) was checked in a separate if statement right after L463